kotlintest-core/src/commonMain/kotlin/io/kotlintest/core/specs/AbstractSpecDsl.kt

```diff
@@ -0,0 +1,106 @@
+package io.kotlintest.core.specs
+
+import io.kotlintest.AssertionMode
+import io.kotlintest.IsolationMode
+import io.kotlintest.Spec
+import io.kotlintest.Tag
+import io.kotlintest.TestCase
+import io.kotlintest.TestCaseOrder
+import io.kotlintest.TestResult
+import io.kotlintest.extensions.SpecLevelExtension
+import io.kotlintest.extensions.TestListener
+
+abstract class AbstractSpecDsl : AbstractSpec() {
+
+   private var beforeTestFn: (TestCase) -> Unit = {}
+   private var afterTestFn: (TestCase, TestResult) -> Unit = { _, _ -> }
+   private var beforeSpecFn: (Spec) -> Unit = {}
+   private var afterSpecFn: (Spec) -> Unit = {}
+   private var afterSpecClassFn: (Spec, Map<TestCase, TestResult>) -> Unit = { _, _ -> }
+   private var tags: Set<Tag> = emptySet()
+   private var listeners: List<TestListener> = emptyList()
+   private var extensions: List<SpecLevelExtension> = emptyList()
+   private var testCaseOrder: TestCaseOrder? = null
+   private var isolationMode: IsolationMode? = null
+   private var assertionMode: AssertionMode? = null
+
+   override fun testCaseOrder(): TestCaseOrder? = testCaseOrder
+   override fun isolationMode(): IsolationMode? = isolationMode
+   override fun assertionMode(): AssertionMode? = assertionMode
+
+   override fun listeners(): List<TestListener> = listeners
+   override fun extensions(): List<SpecLevelExtension> = extensions
+
+   override fun beforeTest(testCase: TestCase) {
+      super.beforeTest(testCase)
+      beforeTestFn(testCase)
+   }
+
+   override fun afterTest(testCase: TestCase, result: TestResult) {
+      super.afterTest(testCase, result)
+      afterTestFn(testCase, result)
+   }
+
+   override fun beforeSpec(spec: Spec) {
+      super.beforeSpec(spec)
+      beforeSpecFn(spec)
+   }
+
+   override fun afterSpec(spec: Spec) {
+      super.afterSpec(spec)
+      afterSpecFn(spec)
+   }
+
+   override fun tags(): Set<Tag> {
+      return super.tags()
+   }
+
+   override fun afterSpecClass(spec: Spec, results: Map<TestCase, TestResult>) {
+      super.afterSpecClass(spec, results)
+      afterSpecClassFn(spec, results)
+   }
+
+   fun set(isolationMode: IsolationMode) {
+      this.isolationMode = isolationMode
+   }
+
+   fun set(assertionMode: AssertionMode) {
+      this.assertionMode = assertionMode
+   }
+
+   fun set(testCaseOrder: TestCaseOrder) {
+      this.testCaseOrder = testCaseOrder
+   }
+
+   fun tags(vararg tags: Tag) {
+      this.tags = tags.toSet()
```

Comment:
Do we think that perhaps we want Tags, Listeners and Extensions to be called more than once, and this could be an increment instead of an asignment?
`this.tags += tags.toSet()`. I don't have a concrete use case for this in my mind yet, but I'm thinking that perhaps when we we want to compose different pieces of code (we don't support that yet, I don't remember if it's proposed), we might want to compose this too.

It wouldn't make a difference for our main use case with this DSL

Reply:
We could have tags that replaces it, and addTag, withTag or just tag to add ?